### ISSUE TITLE
Latest version requires unsafe-inline due to inline styles

### ISSUE THREAD

**Issue by rupertbg** [xtermjs/xterm.js#4445]
Content Security Policies need to be set to 'unsafe-inline' to work with xterm.js. Older versions didn't use inline styles so this wasn't an issue.

Ideally xterm should stop using inline styles or support a user-provided nonce value that can be set in the CSP. See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src

## Details
- Browser and browser version: all
- OS version: all
- xterm.js version: 5.1.0

### Steps to reproduce

1. Set a content security policy like "style-src 'self';"
2. Make an xterm that has a resizable container
3. Resizing causes CSP errors in the console.
4. Resizing doesn't work properly

**Comment by Tyriar** [comment]
It always needed it depending on the renderer you are using. Is the proposal that the nonce is generated by the server and passed into xterm.js as an option?

**Comment by Codehardt** [comment]
Same issue here, properties like font-family are blocked by CSP in newest version 5.2.1. I would be fine if I can pass a server-generated nonce to the Terminal constructor.

Kind regards,
Marcel

**Comment by jerch** [comment]
What would be the best way forward here? As I have lit. no clue about CSP scoping rules, some help from more experienced CSP-focused devs would be appreciated.

- How does CSP interact with JS inline `.style` access? Should we rework all those to explicit style elements with a nounce passed through? (That would be quite a big refactoring...)
- Ideally we find the most restrictive still working set of CSP directives, and rework the code accordingly.
- In the end we prolly should create some docs about CSP usage with xtermjs and the catches.

**Comment by rupertbg** [comment]
Well one option would be to use stylesheets instead of inline styles, which is the reason it became a problem in the first place.

**Comment by jerch** [comment]
> Well one option would be to use stylesheets instead of inline styles, which is the reason it became a problem in the first place.

Thats not possible for all styling parts, as some are calculated from metrics calculated in JS.

**Comment by rupertbg** [comment]
> > Well one option would be to use stylesheets instead of inline styles, which is the reason it became a problem in the first place.
> 
> Thats not possible for all styling parts, as some are calculated from metrics in JS.

You could consider updating css variables via `setProperty` on the `:root` instead of applying inline styles. I believe this does not count for the purposes of CSP.

**Comment by Codehardt** [comment]
The only thing that CSP is blocking if you use inline `<style>...</style>` or inline `style="..."`. But there are still ways to dynamically set styles with JS that are not blocked by CSP, by using `elem.setProperty(<prop>, <value>)` or `elem.style.<prop> = <value>`. If this is no option, then there is also a nonce possible: Just allow passing a nonce string in the Terminal constructor and use this nonce in all `<style>` sections: `<style nonce="<the nonce>">...</style>`.

**Comment by Tyriar** [comment]
> But there are still ways to dynamically set styles with JS that are not blocked by CSP, by using elem.setProperty(<prop>, <value>) or elem.style.<prop> = <value>.

This is not an option as styling every single character like this would hurt performance.

> Just allow passing a nonce string in the Terminal constructor and use this nonce in all <style> sections: <style nonce="<the nonce>">...</style>.

This sounds fine, let's add a `nonce`/`cspNonce` option?

**Comment by SimonSiefke** [comment]
Maybe the `CSSStyleSheet` api could useful for this, for example:

```js
const sheet = new CSSStyleSheet();
sheet.replaceSync("a { color: red; }");
document.adoptedStyleSheets.push(sheet)
```

**Comment by Tyriar** [comment]
@SimonSiefke if that works around the CSP issue we could use that method in the browsers that support it 👍

**Comment by Tyriar** [comment]
This change was reverted in https://github.com/xtermjs/xterm.js/pull/4680

**Comment by deadbaed** [comment]
A workaround until CSPs are implemented again: use the [webgl addon](https://github.com/xtermjs/xterm.js/tree/5.5.0/addons/addon-webgl)!

I was getting CSP errors when importing xtermjs in my angular app, and had only a non functional black box. By using the webgl backend I am still getting CSP errors, but now I have a functional xtermjs!

**Comment by ssh-peppe** [comment]
With the 6.0.0 release it seems there is now a CSP warning even with the webgl renderer, and the scrollbars fail to render due to that. This is a regression from 5.x.x where this was working.

**Comment by ssh-peppe** [comment]
The code that triggers this, from the Chrome dev console, is:

```js
this._register(tr.runAndSubscribe(o.onChangeColors, () => {
                    this._styleElement.textContent = [".xterm .xterm-scrollable-element > .scrollbar > .slider {", `  background: ${o.colors.scrollbarSliderBackground.css};`, "}", ".xterm .xterm-scrollable-element > .scrollbar > .slider:hover {", `  background: ${o.colors.scrollbarSliderHoverBackground.css};`, "}", ".xterm .xterm-scrollable-element > .scrollbar > .slider.active {", `  background: ${o.colors.scrollbarSliderActiveBackground.css};`, "}"].join("\n")
                }
```

**Comment by Tyriar** [comment]
@ssh-peppe I don't know much about unsafe-inline, how hard is the nonce thing mentioned above to implement? We don't hit this in VS Code so ideally the fix would be driven by someone who it does impact.

**Comment by ssh-peppe** [comment]
> [@ssh-peppe](https://github.com/ssh-peppe) I don't know much about unsafe-inline, how hard is the nonce thing mentioned above to implement? We don't hit this in VS Code so ideally the fix would be driven by someone who it does impact.

So what happens there is that if you serve xtermjs from a webserver that has a secure content security policy (the Content-Security-Policy http header), then unsafe inline definitions of CSS styles would basically always be disallowed. You can modify styles using the DOM, i.e. foo.style.background = something, but not unject it as text with textContent for a style element.

The hash or nonce work-arounds are not really feasible since we would need to know what the nonce will be when serving the CSP, but the script tag is produced by xterm and could in principle contain absolutely anything. If the nonce is provided by the app using xtermjs, then it should still be different each time it is served, and would require the CSP to be dynamically created, which is not going to be possible in most cases (including for us).

The particular feature should be implemented such that you don't create CSS as text content, but rather by setting style properties using DOM functions. It might also be possible to use static CSS that references CSS variables, and then you simply set the value of those CS variables if you need changes.

So basically you have static CSS in a file like::

```css
.xterm .xterm-scrollable-element > .scrollbar > .slider.active {
  background: var(--xterm-slider-active);
}
```

And then using code you set:

```js
document.documentElement.style.setProperty('--xterm-slider-active', '#aabbcc');
```

The alternative is to change it at runtime with:

```js
sliderElement.style.background = '#aabbcc';
```

But if you want hover effects you'd need to do that when the cursor hovers over the element, and same for activation (clicks) of course.

**Comment by ssh-peppe** [comment]
Btw, just noticed you have .slider.active, but .slider:hover. I wonder if that is a bug and the intended was .slider:active.

**Comment by Tyriar** [comment]
That particular component is pulled in from an older version of vscode's codebase. There's no way to support this in such a way that would allow the use of inline styles? We could add a nonce option to the API for example.

**Comment by ssh-peppe** [comment]
> That particular component is pulled in from an older version of vscode's codebase. There's no way to support this in such a way that would allow the use of inline styles? We could add a nonce option to the API for example.

A nonce would in practice have to be static, i.e. always served with the same value, which is against the principle of being a nonce ("for one occasion"). If a CSP nonce is static, an attacker can inject arbitrary CSS and simply reuse the same nonce, completely bypassing CSP.

The reason it would in practice have to be static is that typically a CSP could be served from something like nginx, from a static configuration file. In our case we're talking about a security product, installed in our customers' premises. So we don't always even have control over what they are running that serves the CSP.

**Comment by Tyriar** [comment]
Makes sense.

If that's the only case, that's actually in xterm.js not the helpers:

https://github.com/xtermjs/xterm.js/blob/8112fbf21051938530491fc298fd8ffca56978c8/src/browser/Viewport.ts#L84-L96

So moving to CSS variables sounds good to me.

Could also do the same for the DOM renderer to properly support that too:

https://github.com/xtermjs/xterm.js/blob/8112fbf21051938530491fc298fd8ffca56978c8/src/browser/renderer/dom/DomRenderer.ts#L163-L166

**Comment by dolonet** [comment]
Put up #5805 which handles the two sites linked above — dropping the scrollbar-slider `<style>.textContent` in `Viewport.ts` and the static row-span rule in `DomRenderer.ts`. The scrollbar colors now flow through CSS custom properties set on the terminal element via `element.style.setProperty(...)`, which isn't subject to `style-src`, and the row-span rule moves into `xterm.css`. Multiple terminals on one page stay isolated because the variables are scoped to each terminal's root, not `:root`.

That's enough to unblock people running under `style-src 'self'` for the common render path. `DomRenderer._injectCss` still builds the per-terminal theme stylesheet as `<style>` textContent — the ANSI color classes and keyframes in there would need a bigger refactor, so I left that for a follow-up rather than bundling it into this PR.

Hit this when wiring xterm.js into [websh](https://github.com/dolonet/websh) — a little browser-based SSH terminal we host behind a fairly strict CSP. It's been a nice lightweight stack for anyone wanting a web SSH shell without pulling in a heavyweight framework, and fixing the CSP issue upstream is obviously the better long-term answer than loosening the policy.

**Comment by CedricDuffournetInfomaniak** [comment]
Are there any workarounds other than downgrading the package to 5.5

**Comment by dporf** [comment]
@dolonet 
> [dolonet](https://github.com/dolonet) closed this by deleting the head repository [on Apr 13](https://github.com/xtermjs/xterm.js/pull/5805#event-24444392493)

was it intentional?

**Comment by nabramovitz** [comment]
A data point on the constructed-stylesheets direction that was tried and reverted in #4680: the Chromium crash that forced the revert (microsoft/vscode#189753) no longer reproduces in current Chrome.

We tested this while evaluating CSP hardening for Cloud Foundry Stratos (measurements recorded [here](https://github.com/cloudfoundry/stratos/issues/5705#issuecomment-5151952795)). Method: reimplemented the reverted `StyleSheet.ts` API shape verbatim — `new CSSStyleSheet()`, `document.adoptedStyleSheets.push(sheet)`, `sheet.replaceSync(css)` on update, `indexOf` + `splice` on dispose — layered over two live DOM-renderer terminals on `xterm@5.4.0-beta.37`, then drove the crash recipe from the vscode report: two terminals open, dispose the **first** (so its sheets splice out of the middle of `adoptedStyleSheets` while the second stays live), then a 30-cycle resize storm calling `replaceSync` on every fit.

On Chrome 151 (macOS): no crash — the full sequence ran to completion, and with `style-src-elem` carrying no `'unsafe-inline'`, the constructed sheets carried the terminals' entire rendering, since they are exempt from `style-src` altogether. Which is the appeal of this route for the present issue: it needs no nonce API and no per-embedder configuration.

Honest limits of the test: this was a mirror reimplementation of the reverted API over the shipped build, not a build of the reverted branch itself; and the original crash was in Electron (~Chromium 116) — the reverted code carried an `isElectron` guard — so Electron re-verification would still be needed before un-reverting for VS Code's sake. Firefox and Safari were not in the probe.

Given three years of Chromium releases since the revert, the constructed-stylesheets path may be worth re-testing on the VS Code side — it would resolve this issue for browser embedders without any new API surface.
